File: zig/zap/src/skynet.zig
// The skynet benchmark as described here:
// https://github.com/atemerev/skynet
//
// Canonical (TooManyCooks) implementation:
// ../../../cpp/TooManyCooks/skynet.cpp
//
// Each node forks 10 children until depth_max levels deep (10^depth_max =
// 100M leaf tasks) and sums their results. zap's ThreadPool is schedule-only,
// so each node is a heap-allocated continuation frame; the 10 children are
// submitted as a single Batch (zap's batch-scheduling feature) and completions
// cascade up through atomic pending counts.
const std = @import("std");
const ThreadPool = @import("thread_pool");
const bu = @import("benchutil");

const depth_max = 8;
const iter_count = 1;

var pool: ThreadPool = undefined;
var done: bu.DoneEvent = .{};

const SkyFrame = struct {
    task: ThreadPool.Task = .{ .callback = run },
    base: u64,
    depth: u32,
    parent: ?*SkyFrame,
    result: std.atomic.Value(u64) = std.atomic.Value(u64).init(0),
    pending: std.atomic.Value(u32) = std.atomic.Value(u32).init(0),

    fn run(task: *ThreadPool.Task) void {
        const frame: *SkyFrame = @fieldParentPtr("task", task);
        if (frame.depth == depth_max) {
            _ = frame.result.fetchAdd(frame.base, .monotonic);
            return finish(frame);
        }
        var offset: u64 = 1;
        var i: u32 = 0;
        while (i < depth_max - frame.depth - 1) : (i += 1) {
            offset *= 10;
        }

        var batch = ThreadPool.Batch{};
        var j: u64 = 0;
        while (j < 10) : (j += 1) {
            const child = bu.allocator.create(SkyFrame) catch @panic("oom");
            child.* = .{
                .base = frame.base + offset * j,
                .depth = frame.depth + 1,
                .parent = frame,
            };
            batch.push(ThreadPool.Batch.from(&child.task));
        }
        frame.pending.store(10, .release);
        pool.schedule(batch);
    }

    fn finish(frame_arg: *SkyFrame) void {
        var frame = frame_arg;
        while (true) {
            const parent = frame.parent orelse {
                done.set(); // root (lives on the main thread's stack)
                return;
            };
            const res = frame.result.load(.monotonic);
            bu.allocator.destroy(frame);
            _ = parent.result.fetchAdd(res, .monotonic);
            if (parent.pending.fetchSub(1, .acq_rel) != 1) return;
            frame = parent;
        }
    }
};

fn skynet(expected: u64) void {
    var root = SkyFrame{ .base = 0, .depth = 0, .parent = null };
    done.reset();
    pool.schedule(ThreadPool.Batch.from(&root.task));
    done.wait();
    const count = root.result.load(.monotonic);
    if (count != expected) {
        bu.print("ERROR: wrong result - {d}\n", .{count});
    }
}

pub fn main(init: std.process.Init.Minimal) void {
    const argv = bu.Argv(1).init(init.args);
    const thread_count = argv.int(u32, 0) orelse bu.defaultThreads();
    bu.print("threads: {d}\n", .{thread_count});
    pool = ThreadPool.init(.{ .max_threads = thread_count });

    var leaves: u64 = 1;
    var i: u32 = 0;
    while (i < depth_max) : (i += 1) {
        leaves *= 10;
    }
    const expected = (leaves - 1) * leaves / 2;

    skynet(expected); // warmup

    bu.print("runs:\n", .{});
    const start = bu.nowNs();
    var iter: usize = 0;
    while (iter < iter_count) : (iter += 1) {
        skynet(expected);
    }
    const dur_us = (bu.nowNs() - start) / std.time.ns_per_us;
    bu.print("  - iteration_count: {d}\n", .{iter_count});
    bu.print("    duration: {d} us\n", .{dur_us});
    bu.print("    max_rss: {d} KiB\n", .{bu.peakMemoryUsageKiB()});

    pool.shutdown();
    pool.deinit();
}

File: rust/rayon/src/bin/skynet.rs
// The skynet benchmark as described here:
// https://github.com/atemerev/skynet
//
// Canonical (TooManyCooks) implementation:
// ../../../../cpp/TooManyCooks/skynet.cpp
//
// Each node forks 10 children until DEPTH_MAX levels deep (10^DEPTH_MAX = 100M
// leaf tasks) and sums their results. In rayon the 10-way fork-and-sum is a
// parallel iterator over 0..10: rayon adaptively splits the range across idle
// workers (via `join` internally) and reduces the per-child results with `sum`.

use rayon::prelude::*;
use rayon_bench::{build_pool, default_threads, peak_memory_usage};
use std::time::Instant;

const DEPTH_MAX: u64 = 8;
const ITER_COUNT: usize = 1;

fn skynet_one(base: u64, depth: u64) -> u64 {
    if depth == DEPTH_MAX {
        return base;
    }
    let mut offset = 1u64;
    for _ in 0..(DEPTH_MAX - depth - 1) {
        offset *= 10;
    }

    (0..10u64)
        .into_par_iter()
        .map(|i| skynet_one(base + offset * i, depth + 1))
        .sum()
}

fn skynet(expected: u64) {
    let count = skynet_one(0, 0);
    if count != expected {
        println!("ERROR: wrong result - {}", count);
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let thread_count = if args.len() > 1 {
        args[1].parse().expect("thread count")
    } else {
        default_threads()
    };

    let leaves = 10u64.pow(DEPTH_MAX as u32);
    let expected = (leaves - 1) * leaves / 2;

    let pool = build_pool(thread_count);
    println!("threads: {}", thread_count);

    pool.install(|| {
        skynet(expected); // warmup

        let start = Instant::now();
        for _ in 0..ITER_COUNT {
            skynet(expected);
        }
        let dur = start.elapsed();

        println!("runs:");
        println!("  - iteration_count: {}", ITER_COUNT);
        println!("    duration: {} us", dur.as_micros());
        println!("    max_rss: {} KiB", peak_memory_usage());
    });
}
